src/Algebra/Graph/NonEmpty.hs

```diff
@@ -0,0 +1,563 @@
+{-# LANGUAGE DeriveFunctor, DeriveFoldable, DeriveTraversable #-}
+-----------------------------------------------------------------------------
+-- |
+-- Module     : Algebra.Graph.NonEmpty
+-- Copyright  : (c) Andrey Mokhov 2016-2017
+-- License    : MIT (see the file LICENSE)
+-- Maintainer : andrey.mokhov@gmail.com
+-- Stability  : experimental
+--
+-- __Alga__ is a library for algebraic construction and manipulation of graphs
+-- in Haskell. See <https://github.com/snowleopard/alga-paper this paper> for the
+-- motivation behind the library, the underlying theory, and implementation details.
+--
+-- This module defines the data type 'NonEmptyGraph' and associated algorithms.
+--
+-----------------------------------------------------------------------------
+module Algebra.Graph.NonEmpty (
+    -- * Algebraic data type for graphs
+    NonEmptyGraph (..),
+
+    -- * Basic graph construction primitives
+    vertex, edge, overlay, connect, vertices1, edges1, overlays1, connects1,
+
+    -- * Graph folding
+    foldg1,
+
+    -- * Relations on graphs
+    isSubgraphOf, (===),
+
+    -- * Graph properties
+    size, hasVertex, hasEdge, vertexCount, edgeCount, vertexList1, edgeList,
+    vertexSet, vertexIntSet, edgeSet,
+
+    -- * Standard families of graphs
+    path1, circuit1, clique1, biclique1, star, mesh1, torus1,
+
+    -- * Graph transformation
+    replaceVertex, mergeVertices, splitVertex1, transpose, simplify,
+
+    -- * Graph composition
+    box
+  ) where
+
+import Prelude ()
+import Prelude.Compat
+
+import Control.DeepSeq (NFData (..))
+import Control.Monad.Compat
+import Data.Foldable
+import Data.List.NonEmpty (NonEmpty (..))
+import Data.Maybe
+
+import qualified Algebra.Graph                    as G
+import qualified Algebra.Graph.AdjacencyMap       as AM
+import qualified Algebra.Graph.Class              as C
+import qualified Algebra.Graph.HigherKinded.Class as H
+import qualified Algebra.Graph.IntAdjacencyMap    as IAM
+import qualified Algebra.Graph.Relation           as R
+import qualified Data.IntSet                      as IntSet
+import qualified Data.List.NonEmpty               as NonEmpty
+import qualified Data.Set                         as Set
+
+data NonEmptyGraph a = Vertex a
```

Comment:
worth documenting that non-empty means that there's always at least one vertex, but there might be zero edges!